### DISCUSSION TITLE
MatchStartedChanged with ScoreUpdated acts weird

### DISCUSSION THREAD

**Discussion by ysomad** [markus-wa/demoinfocs-golang#366] [Q&A] [answered]
Hi Markus! I'm setting team members to my struct from event `MatchStart` and updating score on `ScoreUpdated` but some demos does not have `MatchStart` event and I have to use `MatchStartedChanged` so I can collect player steam ids on the start. 

BUT using `MatchStartedChanged` brokes `ScoreUpdated` event somehow and I'm getting 0 score for one of the teams at the end of the match.

Score update log with `MatchStart`:
```sh
1
0
1
0
2
0
3
0
3
1
3
2
4
2
4
3
4
4
4
5
4
6
4
7
4
8
4
9
4
10
4
11
TEAM SWITCH
4
11
4
11
5
11
6
11
6
12
6
13
6
14
6
15
6
16
```

Score update log with `MatchStartedChanged`:
```sh
...
1
0
1
0
2
0
3
0
3
1
3
2
4
2
4
3
4
4
4
5
4
6
4
7
4
8
4
9
4
10
4
11
TEAM SWITCH
4
11
4
11
5
11
6
11
6
12
6
13
6
14
6
15
16
0

```

**Comment by markus-wa** [comment]
Hi @ysomad - can you please include your code?

Thanks! :slightly_smiling_face:

**Reply by ysomad** [reply]
> Hi @ysomad - can you please include your code?
> 
> Thanks! 🙂

Added below

**Comment by ysomad** [comment]
Not sure is it a bug or I don't understand the behavior of the MatchStartedChanged event.

```go
type parser struct {
	demoinfocs.Parser
	match         *match
}

type match struct {
	team1        matchTeam
	team2        matchTeam
}

type matchTeam struct {
	score          int
        side common.Team
	playerSteamIDs []uint64
}

func (p *parser) parse() {
  // or doing the same thing on events.MatchStartedChanged
  p.RegisterEventHandler(func(_ events.MatchStart) {
        gs := p.GameState()
        t := gs.TeamTerrorists()
	p.match.setTeam1(newMatchTeam(t.ClanName(), t.Flag(), t.Team(), t.Members()))

	ct := gs.TeamCounterTerrorists()
	p.match.setTeam2(newMatchTeam(ct.ClanName(), ct.Flag(), ct.Team(), ct.Members()))
  })

  // acts weird if using with MatchStartedChanged event handler
  p.RegisterEventHandler(func(e events.ScoreUpdated) {
	switch e.TeamState.Team() {
	case p.match.team1.side:
		p.match.team1.score = e.NewScore
	case p.match.team2.side:
		p.match.team2.score = e.NewScore
	}
  })
}
```

**Reply by ysomad** [reply]
I have team switch handle on TeamSideSwitch event but didn't include here.

**Reply by markus-wa** [reply]
I will need something I can compile and run. there's a lot of stuff missing here (e.g newMatchTeam, setTeam1/2)

**Reply by markus-wa** [reply]
I suspect it's because TeamState should not be updated unless players swap teams

**Reply by ysomad** [reply]
I created a branch for you, hope it'll help. https://github.com/ssssargsian/uniplay/tree/markus/MatchStartedChanged

parsing starts in internal/app/app.go
parser implemented in internal/replayparser/parser.go in parser.Parse() method

to run the app: `make run`

Demo without MatchStart event (returns ScoreUpdated always is 0 so I'm using MatchStartedChanging but it returns 0 score one of the teams at the end of match):
https://mega.nz/file/lOR3EZoQ#XfpSbob_F97uUIDVRpVOhUmeL4yNBTh-Wa9L1E27irg

**Reply by markus-wa** [reply]
This should fix your problem - basically you don't want to update teams if MatchStarted was already true / if MatchStarted is set to false


```
diff --git a/internal/replayparser/parser.go b/internal/replayparser/parser.go
index 6fefbdd..1c67f12 100644
--- a/internal/replayparser/parser.go
+++ b/internal/replayparser/parser.go
@@ -37,6 +37,10 @@ func (p *parser) Parse() (parseResult, error) {

        // TODO: https://github.com/markus-wa/demoinfocs-golang/discussions/366
        p.RegisterEventHandler(func(e events.MatchStartedChanged) {
+               if e.OldIsStarted || !e.NewIsStarted {
+                       return
+               }
+
                p.setTeams(p.GameState())
        })
```

**Reply by markus-wa** [reply]
btw, most if not all of the mutexes are not necessary (the ones I saw, anyway).

the parser makes sure to dispatch events sequentially

**Reply by ysomad** [reply]
> btw, most if not all of the mutexes are not necessary (the ones I saw, anyway).
> 
> the parser makes sure to dispatch events sequentially

oh, ok. I'll remove them, thanks!

**Reply by ysomad** [reply]
> This should fix your problem - basically you don't want to update teams if MatchStarted was already true / if MatchStarted is set to false
> 
> ```
> diff --git a/internal/replayparser/parser.go b/internal/replayparser/parser.go
> index 6fefbdd..1c67f12 100644
> --- a/internal/replayparser/parser.go
> +++ b/internal/replayparser/parser.go
> @@ -37,6 +37,10 @@ func (p *parser) Parse() (parseResult, error) {
> 
>         // TODO: https://github.com/markus-wa/demoinfocs-golang/discussions/366
>         p.RegisterEventHandler(func(e events.MatchStartedChanged) {
> +               if e.OldIsStarted || !e.NewIsStarted {
> +                       return
> +               }
> +
>                 p.setTeams(p.GameState())
>         })
> ```

yeah, I tested in multiple demos now it's working in all of them! Brilliant, thank you so much, I was sitting on this literally two days straight :)

**Reply by markus-wa** [reply]
great to hear! glad it solved the issue :partying_face: - happy parsing! :smile:

**Comment by ysomad** [comment, marked as answer]
Do not update teams if MatchStarted was already true / if MatchStarted is set to false
```go
p.RegisterEventHandler(func(e events.MatchStartedChanged) {
              if e.OldIsStarted || !e.NewIsStarted {
                       return
             }

                p.setTeams(p.GameState())
        })
```
